programs/bpf_loader/src/syscalls/sysvar.rs

```diff
@@ -156,3 +159,34 @@ declare_builtin_function!(
         )
     }
 );
+
+declare_builtin_function!(
+    /// Get a slice of a Sysvar in-memory representation
+    SyscallGetSysvar,
+    fn rust(
+        invoke_context: &mut InvokeContext,
+        sysvar_id_addr: u64,
+        length: u64,
+        offset: u64,
+        var_addr: u64,
+        _arg5: u64,
+        memory_mapping: &mut MemoryMapping,
+    ) -> Result<u64, Error> {
+        let check_aligned = invoke_context.get_check_aligned();
+        consume_compute_meter(
+            invoke_context,
+            invoke_context
+                .get_compute_budget()
+                .sysvar_base_cost
+                .saturating_add(length),
+        )?;
+
+        let sysvar_id = translate_type::<Pubkey>(memory_mapping, sysvar_id_addr, check_aligned)?;
+        let var = translate_slice_mut::<u8>(memory_mapping, var_addr, length, check_aligned)?;
+
+        let cache = invoke_context.get_sysvar_cache();
+        cache.read_sysvar_into(sysvar_id, length as usize, offset as usize, var)?;
```

Comment:
will get to this in a separate commit tonight because it requires writing tests

Reply:
to be clear, when the spec says that it "returns" 1, does that mean it should return `Ok(1)` rather than `Err(InstructionError::Custom(1))`? i assume the former because you said "graceful" but this isnt perfectly clear to me from the spec. does this mean passing in 0 for the length should return `Ok(0)` and do nothing? what do we return if the output buffer is too small? what about overflow of adding length to offset, even though that shouldnt be possible to hit? can we return errors at all?

Reply:
to be clear, when the spec says that it "returns" 1, does that mean it should return `Ok(1)` rather than `Err(InstructionError::Custom(1))`? i assume the former because you said "graceful." does this mean passing in 0 for the length should return `Ok(0)` and do nothing?

Reply:
wait a second, returning an error doesnt "abort the virtual machine" either does it...? i actually cant tell what the `ThisError` conversion turns them into either... `SyscallError` maybe? does this mean by the spec we should `panic!` on the failure cases...? and then is returning `Err` a "graceful success"......? im very confused by this and dont understand why we dont return errors uniformly including for the "successful failure" cases

Reply:
ok well as best i can tell after making a test program and putting a `?` in the syscall for `Clock::get` the error is transformed into `SyscallError` the the declare macro and aborts computation somewhere i havent been able to locate through the indirection. so its not quite as complicated as i thought. im still skeptical of returning success codes for the soft failures tho because of implementation risks, particularly if safety mechanisms in rust itself result in a hard failure for agave for what would be a soft failure on firedancer

Reply:
lol actually i think its fine, ive been thinking about this forwards (check the hard errors first because theyre more important) when i should have been thinking about this backwards (check the soft errors first because they cant happen accidentally due to rust). the specific thing i was afraid of was "check for possible hard error -> check for possible soft error -> try to do something that causes a hard error despite checks"

Reply:
ive been wrestling with the consensus implications of having soft and hard errors for a few hours, wrote a few comments here, deleted them for being confused and tentative, and left a more actionable comment on the simd

honestly i dislike the multiple kinds of error tho and would be happier if we could just abort on the cases that the spec wants to return codes

Reply:
Those are great points. I think we need to hash it out in the SIMD PR. I'll comment over there and we can circle back here!

Reply:
Circling back from SIMD conversations, the behavior here should be to abort the VM if any constraints are violated, even if a graceful error constraint is also violated.

https://github.com/solana-foundation/solana-improvement-documents/pull/127#discussion_r1571071566

Reply:
just checking, is this comment just to clarify resolution for future readers? the code does this now, you might not be seeing the latest commit if it looks otherwise (eg if you clicked the link to the file above this comment thread)

Reply:
Yes, I'll mark resolved once I take a look, but yeah just for reader to see the conclusion.